test/integration/scheduler_perf/util.go

```diff
@@ -118,34 +115,33 @@ func mustSetupCluster(ctx context.Context, tb testing.TB, config *config.KubeSch
 		var err error
 		config, err = newDefaultComponentConfig()
 		if err != nil {
-			tb.Fatalf("Error creating default component config: %v", err)
+			tCtx.Fatalf("Error creating default component config: %v", err)
 		}
 	}
 
-	client := clientset.NewForConfigOrDie(cfg)
-	dynClient := dynamic.NewForConfigOrDie(cfg)
+	tCtx = ktesting.WithRESTConfig(tCtx, cfg)
 
 	// Not all config options will be effective but only those mostly related with scheduler performance will
 	// be applied to start a scheduler, most of them are defined in `scheduler.schedulerOptions`.
-	_, informerFactory := util.StartScheduler(ctx, client, cfg, config, outOfTreePluginRegistry)
-	util.StartFakePVController(ctx, client, informerFactory)
-	runGC := util.CreateGCController(ctx, tb, *cfg, informerFactory)
-	runNS := util.CreateNamespaceController(ctx, tb, *cfg, informerFactory)
+	_, informerFactory := util.StartScheduler(tCtx, tCtx.Client(), cfg, config, outOfTreePluginRegistry)
+	util.StartFakePVController(tCtx, tCtx.Client(), informerFactory)
+	runGC := util.CreateGCController(tCtx, tCtx, *cfg, informerFactory)
+	runNS := util.CreateNamespaceController(tCtx, tCtx, *cfg, informerFactory)
 
 	runResourceClaimController := func() {}
 	if enabledFeatures[features.DynamicResourceAllocation] {
 		// Testing of DRA with inline resource claims depends on this
 		// controller for creating and removing ResourceClaims.
-		runResourceClaimController = util.CreateResourceClaimController(ctx, tb, client, informerFactory)
+		runResourceClaimController = util.CreateResourceClaimController(tCtx, tCtx, tCtx.Client(), informerFactory)
 	}
 
-	informerFactory.Start(ctx.Done())
-	informerFactory.WaitForCacheSync(ctx.Done())
+	informerFactory.Start(tCtx.Done())
+	informerFactory.WaitForCacheSync(tCtx.Done())
 	go runGC()
 	go runNS()
 	go runResourceClaimController()
 
-	return informerFactory, client, dynClient
+	return informerFactory, tCtx
```

Comment:
I'm thinking can we not return the ctx again here, but this definitely needs some refactoring of TContext.

Reply:
`mustSetupCluster` gets a `TContext` without Kubernetes clients and creates a new `TContext` which has them, so returning that here is okay.

`TContext` instances are intentionally immutable, i.e. you cannot change one instance by changing some of its attributes. That is the same design that `logr.Logger` and `context.Context` are using. I think it make sense and has some nice properties (no race conditions although there is no locking, prevents callees from changing something that is used by their caller, "composing" contexts by wrapping a base context).